bam_data_store/metadata/definitions.py

```diff
@@ -0,0 +1,428 @@
+            # 'DATE': ,
+            'HYPERLINK': str,
+            'INTEGER': int,
+            # 'MATERIAL': ,
+            'MULTILINE_VARCHAR': str,
+            # 'OBJECT': ,
+            'REAL': float,
+            # 'TIMESTAMP': ,
+            'VARCHAR': str,
+            # 'XML': ,
+        }
+        return mapping.get(self, None)
+
+
+class EntityDef(BaseModel):
+    """
+    Abstract base class for all masterdata entity definitions. The entity definitions are immutable properties.
+    This class provides a common interface (with common attributes like `version`, `code` and
+    `description`.) for all entity definitions.
+    """
+
+    version: int = Field(
+        ...,
+        description="""
+        Version of the entity definition. This is an integer that is incremented each time the
+        definition is changed.
+        """,
+    )  # ? is this useful if we end up doing version control?
+
+    code: str = Field(
+        ...,
+        description="""
+        Code string identifying the entity with an openBIS inventory definition. Note that:
+
+        - Must be uppercase and separated by underscores, e.g. `'EXPERIMENTAL_STEP'`.
+        - If the entity is native to openBIS, the code must start with a dollar sign, e.g. `'$NAME'`.
+        - In the case of inheritance, it needs to be separated by dots, e.g. `'WELDING_EQUIPMENT.INSTRUMENT'`.
+        """,
+    )
+
+    description: str = Field(
+        ...,
+        description="""
+        Description of the entity. This is the human-readable text for the object and must be
+        as complete and concise as possible. The German description can be added after the English
+        description separated by a double slash (//), e.g. `'Chemical Substance//Chemische Substanz'`.
+        """,
+    )
+
+    # TODO check ontology_id, ontology_version, ontology_annotation_id, internal (found in the openBIS docu)
+
+    @field_validator('code')
+    @classmethod
+    def validate_code(cls, value: str) -> str:
+        if not value or not re.match(r'^[A-Z_\$\.]+$', value):
+            raise ValueError(
+                '`code` must follow the rules specified in the description: 1) Must be uppercase, '
+                '2) separated by underscores, 3) start with a dollar sign if native to openBIS, '
+                '4) separated by dots if there is inheritance.'
+            )
+        return value
+
+    @field_validator('description')
+    @classmethod
+    def strip_description(cls, value: str) -> str:
+        return value.strip()
+
+
+class BaseObjectTypeDef(EntityDef):
+    """
+    Definition class used for the common fields for `CollectionTypeDef`, `ObjectTypeDef`, and `DataSetType`.
+    It adds the fields of `validation_script`.
+    """
+
+    validation_script: Optional[str] = Field(
+        default=None,
+        description="""
+        Script written in Jython used to validate the object type.
+        """,
+    )  # ? is this truly used?
```

Comment:
Never seen this "BaseObject" creation (check also with Angela). I think it is possible to move this validation_script to "EntityDef", and make "ObjectTypeDef" inherit from EntityDef directly. If not possible, leave it like it is, but just for the validation_script (I think this will not be created directly)

Reply:
I meant, the field `validation_script`.

The reason why there is a `BaseObjectTypeDef` class is that not all other entities contain a `validation_script` field, so if we put it in `EntityDef`, properties, for example, would have this field (and in openBIS, it seems they don't).

In other words, all entities contain `version`, `code` and `description` (that's why they are defined in `EntityDef`), while not all contain `validation_script` (as you can see in the inheritance, only `CollectionTypeDef`, `ObjectTypeDef` and `DataSetTypeDef` do).

Reply:
I deleted now this question, as I know `validation_script` is added in openBIS; but tbh, we are going to handle this differently, instead of using Jython, we can use these classes and pydantic for validating stuff.